exporter/prometheusremotewriteexporter/exporter.go

```diff
@@ -148,69 +150,115 @@ func validateAndSanitizeExternalLabels(cfg *Config) (map[string]string, error) {
+				}
+			} else {
+				// WAL is enabled, persist requests to the WAL for later export
+				if err := prwe.wal.persistToWAL(requests); err != nil {
+					// Log the error and return a permanent consumer error
+					err = consumererror.NewPermanent(err)
+					prwe.settings.Logger.Error("Failed to persist to WAL:", zap.Error(err))
+				}
+			}
+		}(partitions[i])
 	}
+
+	// Wait for all workers to finish processing the requests
+	wg.Wait()
+
 	return nil
 }
 
-// export sends a Snappy-compressed WriteRequest containing TimeSeries to a remote write endpoint in order
-func (prwe *prwExporter) export(ctx context.Context, requests []*prompb.WriteRequest) error {
-	input := make(chan *prompb.WriteRequest, len(requests))
-	for _, request := range requests {
-		input <- request
+// partitionTimeSeries partitions the time series map into N arrays using hashing/fnv.
+func partitionTimeSeries(tsMap map[string]*prompb.TimeSeries, concurrencyLimit int) []map[string]*prompb.TimeSeries {
+	partitions := make([]map[string]*prompb.TimeSeries, concurrencyLimit)
+	for i := 0; i < concurrencyLimit; i++ {
+		partitions[i] = make(map[string]*prompb.TimeSeries)
 	}
-	close(input)
 
-	var wg sync.WaitGroup
+	// FNV-1a is a non-cryptographic hash function.
+	hasher := fnv.New32a()
+
+	for _, ts := range tsMap {
+		// Hash the labels of the time series using FNV-1a.
+		hasher.Reset()
+		hasher.Write([]byte(labelsToString(ts.Labels)))
+		hash := hasher.Sum32()
+
+		// Modulo the hash with the number of partitions to get the partition index.
+		partitionIndex := int(hash % uint32(concurrencyLimit))
+		partitions[partitionIndex][labelsToString(ts.Labels)] = ts
+	}
 
-	concurrencyLimit := int(math.Min(float64(prwe.concurrency), float64(len(requests))))
-	wg.Add(concurrencyLimit) // used to wait for workers to be finished
+	return partitions
+}
 
+func (prwe *prwExporter) export(ctx context.Context, requests []*prompb.WriteRequest) error {
+	// Use a mutex to handle concurrent errors
 	var mu sync.Mutex
 	var errs error
-	// Run concurrencyLimit of workers until there
-	// is no more requests to execute in the input channel.
-	for i := 0; i < concurrencyLimit; i++ {
-		go func() {
-			defer wg.Done()
-			for {
-				select {
-				case <-ctx.Done(): // Check firstly to ensure that the context wasn't cancelled.
-					return
-
-				case request, ok := <-input:
-					if !ok {
-						return
-					}
-					if errExecute := prwe.execute(ctx, request); errExecute != nil {
-						mu.Lock()
-						errs = multierr.Append(errs, consumererror.NewPermanent(errExecute))
-						mu.Unlock()
-					}
-				}
+
+	// Process each WriteRequest sequentially
+	for _, request := range requests {
+		// Check if the context is already cancelled
+		select {
+		case <-ctx.Done():
+			// Even if the context is already cancelled, return the accumulated errors, if any.
+			return errs
+		default:
+			// Process the WriteRequest
+			if errExecute := prwe.execute(ctx, request); errExecute != nil {
+				mu.Lock()
```

Comment:
You need to fix the error handling here:
1. You don't need mutex in the `export` method because the `errs` variable is local and only accessed by a single worker (this was not the case before).
2. You need to properly propagate error in `handleExport`. You are ignoring errors in case any of the workers find errors. what you can do is create an array of errors and each worker will write to a position. After all workers finish, you can check if there are any errors written to the array and wrap them using a `multierr.Append`  . The waiting group will work as a mutex in the end.